lib/Runtime/Library/EngineInterfaceObject.cpp

```diff
@@ -315,32 +315,47 @@ namespace Js
         EngineInterfaceObject_CommonFunctionProlog(function, callInfo);
+#pragma warning(pop)
 
-        if (callInfo.Count >= 2 && JavascriptFunction::Is(args.Values[1]))
+        AssertOrFailFast((callInfo.Count == 3 || callInfo.Count == 4) && JavascriptFunction::Is(args[1]) && JavascriptString::Is(args[2]));
+
+        JavascriptFunction *func = JavascriptFunction::UnsafeFromVar(args[1]);
+        JavascriptString *methodName = JavascriptString::UnsafeFromVar(args[2]);
+
+        func->GetFunctionProxy()->SetIsPublicLibraryCode();
+
+        const char16 *methodNameBuf = methodName->GetSz();
+        charcount_t methodNameLength = methodName->GetLength();
+        const char16 *shortName = wcsrchr(methodNameBuf, _u('.'));
+        charcount_t shortNameOffset = 0;
+        if (shortName != nullptr)
         {
-            JavascriptFunction* func = JavascriptFunction::FromVar(args.Values[1]);
-            func->GetFunctionProxy()->SetIsPublicLibraryCode();
+            shortName++;
+            shortNameOffset = static_cast<charcount_t>(shortName - methodNameBuf);
+        }
 
-            if (callInfo.Count >= 3 && JavascriptString::Is(args.Values[2]))
-            {
-                JavascriptString* customFunctionName = JavascriptString::FromVar(args.Values[2]);
-                // tagPublicFunction("Intl.Collator", Collator); in Intl.js calls TagPublicLibraryCode the expected name is Collator so we need to calculate the offset
-                const char16 * shortName = wcsrchr(customFunctionName->GetString(), _u('.'));
-                uint shortNameOffset = 0;
-                if (shortName != nullptr)
-                {
-                    // JavascriptString length is bounded by uint max
-                    shortName++;
-                    shortNameOffset = static_cast<uint>(shortName - customFunctionName->GetString());
-                }
-                func->GetFunctionProxy()->EnsureDeserialized()->SetDisplayName(customFunctionName->GetString(), customFunctionName->GetLength(), shortNameOffset);
-            }
+        func->GetFunctionProxy()->EnsureDeserialized()->SetDisplayName(methodNameBuf, methodNameLength, shortNameOffset);
 
-            return func;
+        bool creatingConstructor = true;
+        if (callInfo.Count == 4)
+        {
+            AssertOrFailFast(JavascriptBoolean::Is(args[3]));
+            creatingConstructor = JavascriptBoolean::UnsafeFromVar(args[3])->GetValue();
         }
 
-        return scriptContext->GetLibrary()->GetUndefined();
+        if (!creatingConstructor)
+        {
+            FunctionInfo *info = func->GetFunctionInfo();
+            info->SetAttributes((FunctionInfo::Attributes) (info->GetAttributes() | FunctionInfo::Attributes::ErrorOnNew));
+
+            func->SetAttributes(PropertyIds::prototype, PropertyConfigurable);
+            func->DeleteProperty(PropertyIds::prototype, PropertyOperationFlags::PropertyOperation_Force);
```

Comment:
I noticed that DeleteProperty doesn't respect PropertyOperation_Force, at least not the way I thought it should. It makes it so the operation doesn't throw, but it doesn't bypass configurability. I can change that if we all agree that force should bypass any other restrictions.

Reply:
Since this is used by library code only, are there any cases of non-configurable prototype properties? Do you construct lambdas which go through this path?

Just out of caution I would be inclined to leave `Force` as it is, but it does sound like the intention is that it skips those kinds of restrictions, at least for adding properties.

Reply:
We don't _currently_ put any lambdas through this codepath, but theres nothing stopping us from doing so in the future. Additionally, we don't install anything on the prototypes of functions that go through this particular code path, where the prototype gets deleted. Constructor's prototypes get modified, but they don't hit this.